### ISSUE TITLE
[tvp] "Video requires payment and log-in, but log-in is not implemented"

### ISSUE THREAD

**Issue by Displacedpie** [yt-dlp/yt-dlp#6987]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that a **supported** site is broken
- [X] I've verified that I'm running yt-dlp version **2023.03.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Canada

### Provide a description that is worded well enough to be understood

I've been downloading episodes from website: Vod.tvp.pl / https://vod.tvp.pl/seriale,18/polowanie-na-cmy-odcinki,390116

I've been able to get every episode with no issues what so ever, with the exception of Episode #7. https://vod.tvp.pl/seriale,18/polowanie-na-cmy-odcinki,390116/odcinek-7,S01E07,398869

[tvp:vod] Extracting URL: https://vod.tvp.pl/seriale,18/polowanie-na-cmy-odcinki,390116/odcinek-7,S01E07,398869
[tvp:vod] 398869: Downloading JSON metadata
[tvp:embed] Extracting URL: tvp:69230468
[tvp:embed] 69230468: Downloading webpage
ERROR: [tvp:embed] 69230468: Video requires payment and log-in, but log-in is not implemented; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U

(Same error with -vU)

The video works in browser, I've created an account and tried --cookies-from-browser chrome with no luck. 

I've updated twice hoping for a resolution - video has been available since April 30th, 2023. 

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', 'https://vod.tvp.pl/seriale,18/polowanie-na-cmy-odcinki,390116/odcinek-7,S01E07,398869']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version nightly@2023.05.03.204705 [2f07c4c1d] (win_exe)
[debug] Python 3.8.10 (CPython AMD64 64bit) - Windows-10-10.0.19044-SP0 (OpenSSL 1.1.1k  25 Mar 2021)
[debug] exe versions: none
[debug] Optional libraries: Cryptodome-3.17, brotli-1.0.9, certifi-2022.12.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-11.0.2
[debug] Proxy map: {}
[debug] Loaded 1813 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp-nightly-builds/releases/latest
Available version: nightly@2023.05.03.204705, Current version: nightly@2023.05.03.204705
Current Build Hash: b73b1cb4c64dcbd61a084ba9910dca9837cedb7b450c75605ec5d197cdc8ad01
yt-dlp is up to date (nightly@2023.05.03.204705)
[tvp:vod] Extracting URL: https://vod.tvp.pl/seriale,18/polowanie-na-cmy-odcinki,390116/odcinek-7,S01E07,398869
[tvp:vod] 398869: Downloading JSON metadata
[tvp:embed] Extracting URL: tvp:69230468
[tvp:embed] 69230468: Downloading webpage
ERROR: [tvp:embed] 69230468: Video requires payment and log-in, but log-in is not implemented; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "yt_dlp\extractor\common.py", line 694, in extract
  File "yt_dlp\extractor\tvp.py", line 392, in _real_extract
```

**Comment by dirkf** [comment]
The extractor finds the TVP id `69230468` from a metadata AOI URL (the returned data includes a lot of other and tries this URL: https://www.tvp.pl/sess/TVPlayer2/api.php?id=69230468&@method=getTvpConfig&@callback=sasin70 (`callback` is picked randomly from a list). Fetching that URL returns a page with JSON like this
```py
[{'context': 'id',
  'data': None,
  'desc': 'Obiekt wymaga płatności',
  'type': 'Error'}]
```
The browser JS is using some different method, or passing additional headers, which might become apparent from a network trace using the browser's development tools.

**Comment by Displacedpie** [comment]
Thank you for the reply! 

Could I ask for advice on how to run a network trace? 

I'm pretty new to all of this.

**Comment by dirkf** [comment]
Actually, the answer is this: https://vod.tvp.pl/api/products/398869/videos/playlist?platform=BROWSER&videoType=MOVIE
<details><summary>JSON playlist result</summary>

```py
{
  'sources': {
    'HLS': [
      {
        'src': 'https://rsdt-krk201-4.tvp.pl/token/video/vod/69230468/20230504/785384984/X2P8b7AC2hD-POBG5j1lWx35b4kPhUN76BrY8Sx2N6hXDVBtsw6o8GFB1VSqmw5aDiZutk97EpJd64s-0KV8ZM7s59QW7UQvuwJQVNIv84e9mTIiuLMsfDRW0bM3EigHpQLnSybEg7TeQPaG_-uSuPPlpVcGlC5o8r5cNu3lG_Q/video.ism/video-fmp4.m3u8'
      }
    ],
    'DASH': [
      {
        'src': 'https://rsdt-krk201-4.tvp.pl/token/video/vod/69230468/20230504/785384984/X2P8b7AC2hD-POBG5j1lWx35b4kPhUN76BrY8Sx2N6hXDVBtsw6o8GFB1VSqmw5aDiZutk97EpJd64s-0KV8ZM7s59QW7UQvuwJQVNIv84e9mTIiuLMsfDRW0bM3EigHpQLnSybEg7TeQPaG_-uSuPPlpVcGlC5o8r5cNu3lG_Q/video.ism/video.mpd'
      }
    ]
  },
  'stripe': {
    'id': 120545,
    'url': 'https://klatki.tvp.pl/rbig69230468/stripe_{index}.jpg',
    'interval': 2,
    'count': 8,
    'width': 640,
    'height': 360
  },
  'externalVideoSource': true
}
```
</details>

According to https://github.com/ytdl-org/youtube-dl/pull/32028, there is a new scheme at TVP; your problem show seems only to be accessible through that scheme. Instead of passing the external ID (`69230468`) it now uses the ID `398869` from the URL. The `videoType=MOVIE` is required: I have no idea if other values exist.

Once the yt-dl PR is merged (or ready), it'll come here.

**Comment by gamer191** [comment]
> Could I ask for advice on how to run a network trace?

This issue seems to have already been solved, but if you want to learn reverse engineering, I’d recommend playing with your browser’s dev tools (ctrl-shift-i). The network tab is the most useful, since it lets you see what network requests the webpage is making (in other words, it lets you see where the website is getting the video from)

**Comment by october262** [comment]
this video - https://vod.tvp.pl/seriale,18/polowanie-na-cmy-odcinki,390116/odcinek-7,S01E07,398869 can be downloaded & played without payment.
i used brave browser & the browser add-on  called  "the stream detector"
set the stream detector to ffmpeg, now choose the option to watch with ads,
copy the mpd & download.

**Comment by Gwynnbleid1** [comment]
I tried to check if this fix would let me download a movie (https://vod.tvp.pl/filmy-fabularne,136/masz-talent,386831) from the same site. 
I have updated my binary to nigtly and got: 

`.\yt-dlp.exe "https://vod.tvp.pl/filmy-fabularne,136/masz-talent,386831"
[tvp:vod] Extracting URL: https://vod.tvp.pl/filmy-fabularne,136/masz-talent,386831
[tvp:vod] 386831: Downloading JSON metadata
ERROR: '_io.BytesIO' object has no attribute 'status'`

I got the same result with nightly@2023.05.06.001010. 

With nightly@2023.05.05.055923 I received: 
`.\yt-dlp.exe "https://vod.tvp.pl/filmy-fabularne,136/masz-talent,386831"
[tvp:vod] Extracting URL: https://vod.tvp.pl/filmy-fabularne,136/masz-talent,386831
[tvp:vod] 386831: Downloading JSON metadata
[tvp:embed] Extracting URL: tvp:33745643
[tvp:embed] 33745643: Downloading webpage
ERROR: [tvp:embed] 33745643: Video requires payment and log-in, but log-in is not implemented; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U`

Here is result of -vU command: 

`.\yt-dlp.exe -vU "https://vod.tvp.pl/filmy-fabularne,136/masz-talent,386831"
[debug] Command-line config: ['-vU', 'https://vod.tvp.pl/filmy-fabularne,136/masz-talent,386831']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version nightly@2023.05.05.055923 [ddae33754] (win_exe)
[debug] Python 3.8.10 (CPython AMD64 64bit) - Windows-10-10.0.19045-SP0 (OpenSSL 1.1.1k  25 Mar 2021)
[debug] exe versions: ffmpeg 2023-01-04-git-4a80db5fc2-full_build-www.gyan.dev (setts), ffprobe 2023-01-04-git-4a80db5fc2-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome-3.17, brotli-1.0.9, certifi-2022.12.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-11.0.2
[debug] Proxy map: {}
[debug] Loaded 1813 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp-nightly-builds/releases/latest
Available version: nightly@2023.06.13.032334, Current version: nightly@2023.05.05.055923
Current Build Hash: 2cab78434aaa749ed9a8a2bb87585fd233a33a8940f521efcc26a4b2c5e55d0f
[debug] Downloading _update_spec from https://github.com/yt-dlp/yt-dlp-nightly-builds/releases/latest/download/_update_spec
Updating to nightly@2023.06.13.032334 ...
[debug] Downloading yt-dlp.exe from https://github.com/yt-dlp/yt-dlp-nightly-builds/releases/latest/download/yt-dlp.exe
[debug] Downloading SHA2-256SUMS from https://github.com/yt-dlp/yt-dlp-nightly-builds/releases/latest/download/SHA2-256SUMS
Updated yt-dlp to nightly@2023.06.13.032334
[debug] Restarting: "C:\Users\Matys\Downloads\ffmpeg-2023-01-04-git-4a80db5fc2-full_build\bin\yt-dlp.exe" -vU "https://vod.tvp.pl/filmy-fabularne,136/masz-talent,386831"
[debug] Command-line config: ['-vU', 'https://vod.tvp.pl/filmy-fabularne,136/masz-talent,386831']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version nightly@2023.06.13.032334 [cab94a0cd] (win_exe)
[debug] Python 3.8.10 (CPython AMD64 64bit) - Windows-10-10.0.19045-SP0 (OpenSSL 1.1.1k  25 Mar 2021)
[debug] exe versions: ffmpeg 2023-01-04-git-4a80db5fc2-full_build-www.gyan.dev (setts), ffprobe 2023-01-04-git-4a80db5fc2-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome-3.18.0, brotli-1.0.9, certifi-2023.05.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-11.0.3
[debug] Proxy map: {}
[debug] Loaded 1843 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp-nightly-builds/releases/latest
Available version: nightly@2023.06.13.032334, Current version: nightly@2023.06.13.032334
Current Build Hash: 1e2a38e9ddb7199bf1fc977c322e9fbffd4daca6eb77e05b840fdb619306c623
yt-dlp is up to date (nightly@2023.06.13.032334)
[tvp:vod] Extracting URL: https://vod.tvp.pl/filmy-fabularne,136/masz-talent,386831
[tvp:vod] 386831: Downloading JSON metadata
ERROR: '_io.BytesIO' object has no attribute 'status'
Traceback (most recent call last):
  File "yt_dlp\YoutubeDL.py", line 1552, in wrapper
  File "yt_dlp\YoutubeDL.py", line 1628, in __extract_info
  File "yt_dlp\extractor\common.py", line 704, in extract
  File "yt_dlp\extractor\tvp.py", line 568, in _real_extract
  File "yt_dlp\extractor\tvp.py", line 491, in _call_api
  File "tempfile.py", line 468, in __getattr__
AttributeError: '_io.BytesIO' object has no attribute 'status'`

Am I doing something wrong or should I post a new issue releated to this ?

**Comment by dirkf** [comment]
In this fragment from the updated extractor code (l.485-), the `status` attribute isn't supported in Py < 3.9:
```py
    def _call_api(self, resource, video_id, query={}, **kwargs):
        is_valid = lambda x: 200 <= x < 300
        document, urlh = self._download_json_handle(
            f'{self._API_BASE_URL}/{resource}', video_id,
            query={'lang': 'pl', 'platform': 'BROWSER', **query},
            expected_status=lambda x: is_valid(x) or 400 <= x < 500, **kwargs)
        if is_valid(urlh.status):
            return document
        raise ExtractorError(f'Woronicza said: {document.get("code")} (HTTP {urlh.status})')
```
For the claimed 3.7+ support, something like this is needed in a compat module
```py
import urllib.response
try:
    urllib.response.addinfourl.status
except AttributeError:
    urllib.response.addinfourl.status = property(lambda self: self.getcode())
```

**Comment by coletdjnz** [comment]
All extractors that are using `.status` need to be updated to use `getcode()` for now.

No point adding compat support in the backend - we're merging a rewrite soon which adds our own response interface.
